### PR TITLE
Fix OSV mit working with PURLs

### PR DESCRIPTION
**Pull request by sschuberth**
Please have a look at the individual commit messages for the details.

### CODE REVIEW

File: clients/osv/src/main/kotlin/OsvService.kt [hunk 1]
```diff
@@ -67,7 +67,9 @@ class OsvService(serverUrl: String? = null, httpClient: OkHttpClient? = null) {
             val response = client.getVulnerabilityIdsForPackages(batchRequest).execute()
             val body = response.body()
 
-            if (!response.isSuccessful || body == null) return Result.failure(IOException(response.message()))
+            if (!response.isSuccessful || body == null) {
```
Comment (fviernau) [inline clients/osv/src/main/kotlin/OsvService.kt:70]: Why is `|| body == null` necessary and / or correct?

File: advisor/src/funTest/kotlin/OsvFunTest.kt [hunk 2]
```diff
@@ -77,7 +78,7 @@ class OsvFunTest : StringSpec({
 })
 
 private fun identifierToPackage(id: String): Package =
-    Package.EMPTY.copy(id = Identifier(id))
+    Identifier(id).let { Package.EMPTY.copy(id = it, purl = it.toPurl()) }
```
Comment (fviernau) [inline advisor/src/funTest/kotlin/OsvFunTest.kt:81]: Would swapping the order of this and the following commit lead to the situation that all tests pass of all of the three commits of this PR? If so, would it make sense to change the order of the top-most two commits?

**Review by fviernau — CHANGES_REQUESTED** [review]

File: advisor/src/funTest/kotlin/OsvFunTest.kt (hunk 2 above)
Comment (sschuberth) [reply advisor/src/funTest/kotlin/OsvFunTest.kt:81]: Yes, that would be the case, but I deliberately did not do so as I found it strange to commit a fixup for a commit that I reverted before. That could raise the question whether the fixup should have also been reverted.

File: clients/osv/src/main/kotlin/OsvService.kt (hunk 1 above)
Comment (sschuberth) [reply clients/osv/src/main/kotlin/OsvService.kt:70]: While theoretically (i.e. on HTTP specification level) a successful response could return a null body, in the context of OSV we always expect to get a (non-nullable) `VulnerabilitiesForPackageBatchResponse`, so handling a `null` body as an error seems to be correct.

In any case, the code was like that before and is out of scope for this change.

But I'll make another change to improve error reporting via `errorBody()`.

File: advisor/src/main/kotlin/advisors/Osv.kt [hunk 3]
```diff
@@ -154,6 +154,7 @@ private fun createRequest(pkg: Package): VulnerabilitiesForPackageRequest? {
 
     if (name.isNotBlank() && pkg.id.version.isNotBlank() && !ecosystem.isNullOrBlank()) {
         return VulnerabilitiesForPackageRequest(
+            // Do not specify the purl here as it is mutually exclusive with the ecosystem.
```
Comment (fviernau) [inline advisor/src/main/kotlin/advisors/Osv.kt:157]: Consider making this comment also explain why it is preferable to use ecosystem.

**Review by fviernau — APPROVED** [review]

**Pull request merged**